# Tables

<br />

## Samples filtering

<div align="justify">For each of the nine cucurbitaceae analyzed, a table has been created showing the number of reads after trimming and the type of library for each of the downloaded samples. In addition, information is also provided on which samples have passed both steps.</div><br />

## LncRNAs

<div align="justify">For each of the nine cucurbitaceae analyzed, four folders have been created:</div><br />

 - <div align="justify"><b>Database:</b> TSV file with the non-redundant lncRNAs predicted in this study. Presents detailed information about genomic location, certain molecular characteristics, coding capacity, annotation with housekeeping RNAs and miRNA precursors, annotation with databases of known potential lncRNAs and confidence level.</div>
 - <div align="justify"><b>Fasta:</b> FASTA file with the non-redundant lncRNAs predicted in this study.</div>
 - <div align="justify"><b>Gtf:</b> GTF annotation file with the non-redundant lncRNAs predicted in this study.</div>
 - <div align="justify"><b>Table:</b> TSV file with the non-redundant lncRNAs predicted in this study. It presents less information than the TSV file in the Database folder. </div>

<br />

## PCGs

 <div align="justify">For each of the nine cucurbitaceae analyzed, three folders have been created:</div><br />
  
 - <div align="justify"><b>Fasta:</b> FASTA file with the PCGs.</div>
 - <div align="justify"><b>Gtf:</b> GTF annotation file with the PCGs.</div>
 - <div align="justify"><b>Table:</b> TSV file with the PCGs.</div>

<br />

<div align="center"> <sub><b>Table: Detailed information about all RNA-Seq samples used in this study is available in Supplementary Data 1. in the nine selected cucurbit species.</b></sub> </div>

<br />

| <sub>Species</sub>                        | <sub>Genome Version</sub>    | <sub>Database</sub>           |
|-------------------------------------------|-------------------|--------------------|
| <sub>*Cucumis sativus* (csa)</sub>        | <sub>Chinese Long v.3</sub>  | <sub>Cucurbit Genomics</sub>  |
| <sub>*Cucumis melo* (cme)</sub>           | <sub>DHL92 v.4</sub>         | <sub>Melonomics</sub>         |
| <sub>*Citrullus lanatus* (cla)</sub>      | <sub>97103 v.2</sub>         | <sub>Cucurbit Genomics</sub>  |
| <sub>*Lagenaria siceraria* (lsi)</sub>    | <sub>USVL1VR-Ls v.1</sub>    | <sub>Cucurbit Genomics</sub>  |
| <sub>*Cucurbita moschata* (cmo)</sub>     | <sub>v.1</sub>               | <sub>Cucurbit Genomics</sub>  |
| <sub>*Cucurbita argyrosperma* (car)</sub> | <sub>v.2</sub>               | <sub>Cucurbit Genomics</sub>  |
| <sub>*Cucurbita pepo* (cpe)</sub>         | <sub>v.4.1</sub>             | <sub>Cucurbit Genomics</sub>  |
| <sub>*Cucurbita maxima* (cma)</sub>       | <sub>v1.1</sub>              | <sub>Cucurbit Genomics</sub>  |
| <sub>*Momordica charantia* (mch)</sub>    | <sub>Dali-11 v.1</sub>       | <sub>Cucurbit Genomics</sub>  |

<br />

## Final_summary

<div align="justify">For each of the nine cucurbits analyzed, a table has been created showing all the information obtained in the paper for each of the predicted non-redundant lncRNAs.</div><br />

## Authors

* **Pascual Villalba-Bermell** - *Initial work* - [pasviber](https://github.com/pasviber) (pascual.villalba@csic.es) at [ncRNA-lab](https://github.com/ncRNA-lab).<br />

## Citation

* <div align="justify"> Villalba-Bermell P., Marquez-Molins J. and Gomez G. Identification, characterization and transcriptional analysis of long non-coding RNAs in Cucurbits. BioRxiv [Preprint]. January 15, 2024. Available from: https://doi.org/10.1101/2024.01.12.575433. </div>

