## Usage

<br />

The pipeline execution process using the Cucumis melo species is shown below.

<br />

### STEP 0: Select RNA-seq samples

<div align="justify"> We select all RNA-seq samples from Sequence Read Archive database (SRA) and from a particular species. To this end, we use the <a href="https://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/">E-utilities</a>. <br /><br /></div>

```
cd 01-Sample_processing_and_selection/cme/

sbatch 00-Samples.sh
```

<div align="justify"> As a result, we obtain a list of RNA-seq sample accessions and a metadata table with all the information by RNA-seq sample accession.</div>

<br />


### STEP 1: Download RNA-seq samples

<div align="justify"> All raw RNA-seq samples are downloaded from Sequence Read Archive database (SRA) using <a href="https://github.com/ncbi/sra-tools">SRA Toolkit</a>. <br /><br /></div>

```
cd 01-Sample_processing_and_selection/cme/

sbatch 01-Download.sh
```

<div align="justify"> <a href="https://github.com/s-andrews/FastQC">FastQC</a> and <a href="https://github.com/MultiQC/MultiQC">Multiqc</a> are used to provide some quality control checks on raw sequence data.</div>

<br />


### STEP 2: Quality control RNA-seq samples

<div align="justify">To get clean RNA-seq samples, we remove adapters and filter raw RNA-seq samples by quality using <a href="https://github.com/OpenGene/fastp">Fastp</a>. <br /><br /></div>

```
cd 01-Sample_processing_and_selection/cme/

sbatch 02-Trimming.sh
```

<div align="justify"> To eliminate possible adapters, fasta files containing adapters from the <a href="https://github.com/usadellab/Trimmomatic/tree/main/adapters">Trimmomatic</a> program are used, but other fasta files can be used. It is even possible to gather all adapters in the same file and use a single fasta file. </div>

<div align="justify"> <br /><a href="https://github.com/s-andrews/FastQC">FastQC</a> and <a href="https://github.com/MultiQC/MultiQC">Multiqc</a> are used to provide some quality control checks on clean sequence data.</div>

<div align="justify"> <br />As a summary, a table is generated with the library size for each sample. </div>

<br />


### STEP 3: Select strand-specific RNA-seq samples

<div align="justify"> We deduce whether the RNA-seq samples are strand-specific or not, and select the strand-specific samples using <a href="https://github.com/COMBINE-lab/salmon">Salmon</a>. As Salmon requires the transcriptome file (FASTA) to pseudomap reads, we use <a href="https://github.com/deweylab/RSEM">RSEM</a> to extract this information from the annotation file (GTF) and the genome file (FASTA). <br /><br /></div>

```
cd 01-Sample_processing_and_selection/cme/

sbatch 03-Select_strand-specific_libraries.sh
```

<div align="justify"> As a summary, a table is generated with the library type for each sample. </div>

<br />


### STEP 4: Mapping

<div align="justify"> We map the clean data to the reference genome. To this end, we use <a href="https://github.com/DaehwanKimLab/hisat2">HISAT2</a>. In addition, we also use <a href="https://github.com/biod/sambamba">Sambamba</a> to convert the sam file to bam file and sort it. <br /><br /></div>

```
cd 02-Mapping/cme/

sbatch 04-Mapping.sh
```

<div align="justify"> The use of a different number of threads in the HISAT2 program may affect the mapping percentage. However, this variation is very small and has an insignificant effect on the final assembly. As this is a pipeline under development, a change of the mapping tool is considered in the future if this problem is not solved. </div>

<br />


### STEP 5: Assembly

<div align="justify"> Bam files generated and sorted in the previous step are assembled using <a href="https://github.com/gpertea/stringtie">StringTie2</a>. <br /><br /></div>

```
cd 03-Assembly/cme/

sbatch 05-Assembly.sh
```

<div align="justify"> As a result, an assembly file (GTF) is generated by specie. </div>

<br />


### STEP 6: Transcript annotation

<div align="justify"> Assembled transcripts are classified by their genomic position relative to the protein-coding genes (PCGs) using <a href="https://github.com/gpertea/gffcompare">GffCompare</a>. <br /><br /></div>

```
cd 04-Transcript_annotation/cme/

sbatch 06-Transcript_annotation.sh
```

<div align="justify"> As a result, a class code is assigned to each transcript. </div>

<br />


### STEP 7: LncRNA prediction

<div align="justify"> Assembled and positionally annotated transcripts (GTF) are used to predict potential lncRNAs using several programs and databases (more details in the paper). <br /><br /></div>

```
cd 05-LncRNAs_prediction/cme/

sbatch 07-LncRNAs_prediction.sh
```

<div align="justify"> As a result, we obtain a potential lncRNA database in FASTA, GTF and TSV format. </div>

<br />


### STEP 8: Quantification

<div align="justify"> We quantify the potential lncRNAs, PCGs and both categories (potential lncRNAs and PCGs) using <a href="https://github.com/COMBINE-lab/salmon">Salmon</a>. As Salmon requires the transcriptome file (FASTA) to quantify the transcripts, we use <a href="https://github.com/deweylab/RSEM">RSEM</a> to extract this information from the annotation file (GTF) and the genome file (FASTA). Three different transcriptomic references are used: potential lncRNAs only, PCGs only and both categories together. <br /><br /></div>

```
cd 06-Quantification/cme/

sbatch 08-Quantification_LncRNAs.sh
sbatch 08-Quantification_Genes.sh
sbatch 08-Quantification_ALL.sh
```

<div align="justify"> As a result, three TPM tables (potential lncRNAs only, PCGs only and both categories together) encompassing all samples are generated. </div>

<br />


### STEP 9: Get random intergenic regions

<div align="justify"> We generate a third category of sequences corresponding to random intergenic genome regions of 500 nucleotides that donâ€™t match any of the other two categories (potential lncRNAs and PCGs) using <a href="https://bedtools.readthedocs.io/en/latest/">Bedtools</a>. <br /><br /></div>

```
cd 07-Get_intergenic_regions/cme/

sbatch 09-Random_intergenic_regions.sh
```

<div align="justify"> As a result, we obtain a random intergenic genome regions dataset in FASTA and BED format. </div>

<br />


### STEP 10: Search for TEs and genomic repeats

<div align="justify"> We use the <a href="https://www.repeatmasker.org/RepeatModeler/">RepeatModeler</a> and <a href="https://www.repeatmasker.org/RepeatMasker/">RepeatMasker</a> programs to obtain the repeat content data that can be found in potential lncRNAs, PCGs and random intergenic regions. <br /><br /></div>

```
cd 08-TEs_and_genomic_repeats/cme/

sbatch 10.1-RepeatModeler.sh
sbatch 10.2-RepeatMasker.sh
sbatch 10.3-Intersection_with_PCGs_LncRNAs_and_IR.sh
```

<div align="justify"> As a result, we obtain two tables in TSV format showing the percentage of bases covered by repetitive elements (transposons and other repeat elements) in each category (potential lncRNAs, PCGs and random intergenic regions), distinguishing between the different types of repetitive elements (table 1) or encompassing all of them as a single group (table 2). </div>

<br />


### STEP 11: Molecular properties comparison

<div align="justify"> A common downstream analysis is to compare basic features that differentiate lncRNAs from PCGs such as GC content, number of exons, length, expression level and repeat content. <br /><br /></div>

```
cd 09-Comparison_PCGs_and_lncRNAs/cme/

sbatch 11-Comparison_PCGs_and_lncRNAs.sh
```
<br />


### STEP 12: Genomic distribution

<div align="justify"> The genomic distribution of potential lncRNAs and PCGs is analyzed using <a href="https://bedtools.readthedocs.io/en/latest/">Bedtools</a> and the circlize R package.<br /><br /></div>

```
cd 10-Genomic_distribution/cme/

sbatch 12.1-Density_distribution.sh
sbatch 12.2-Coverage_statistics.sh
```
<br />


### STEP 13: Comparative genomics

<div align="justify"> Evolutionary relationships between species are analyzed at sequence level, positional level (synteny) and motif level. For more details read the paper. <br /><br /></div>

**Sequence level**

```
cd 11-Comparative_genomics/Sequence_level/

sbatch 13-Sequence_level_analysis.sh
```

**Positional level**

<div align="justify"> The approach used to analyze the conservation at positional level is developed and validated in <a href="https://www.tandfonline.com/doi/full/10.1080/15476286.2019.1572438">Pegueroles et al. (2019)</a>. <br /><br /></div>

```
cd 11-Comparative_genomics/Positional_level/

sbatch 13.1-LncRNA_and_PCG_ids_list.sh
sbatch 13.2-Orthologs_identification.sh
sbatch 13.3-Find_syntenic_lncRNAs.sh
sbatch 13.4-Classify_lncRNAs_into_families.sh
sbatch 13.5-Create_figures_and_tables.sh
```

**Motif level**

```
cd 11-Comparative_genomics/Motif_level/

sbatch 13.1-Select_lncRNAs.sh
sbatch 13.2-Prepare_positional_conserved_family_files.sh
sbatch 13.3-Motif_finder-MEME.sh
sbatch 13.4-Motif_enrichment-GOMO.sh
sbatch 13.5-Create_summary_tables.sh
sbatch 13.6-Create_figures_and_tables.sh
```
<br />


### STEP 14: Tissue-specificity analysis

<div align="justify"> Tissue-specificity analysis is performed by <a href="https://apcamargo.github.io/tspex/">Tspex</a> Python package. For more details read the paper. <br /><br /></div>

```
cd 12-Tissue-specificity/cme/

sbatch 14-Tissue-specificity_analysis.sh
```
<br />


### STEP 15: Differential expression analysis

<div align="justify"> Differential expression analysis is performed using DESeq2 and Tximport R packages. For more details read the paper. <br /><br /></div>

```
cd 13-DEA/cme/

sbatch 15-DEA.sh
```
<br />


### STEP 16: Expression correlation analysis

<div align="justify"> Expression correlation analysis is performed using <a href="https://bedtools.readthedocs.io/en/latest/">Bedtools</a> to get the pairs and, DESeq2 and Tximport R packages to get the correlations. For more details read the paper. <br /><br /></div>

**NAT-lncRNAs**

```
cd 14-Expression_correlation/cme/

sbatch 16.1-Antisense-Get_pairs.sh
sbatch 16.2-Antisense-Create_cis_interaction_table.sh
sbatch 16.3-Antisense-Create_cis_correlation_table.sh
sbatch 16.4-Antisense-Join_cis_correlation_tables.sh
sbatch 16.5-Antisense-Create_cis_correlation_table_random.sh
sbatch 16.6-Antisense-Join_cis_correlation_tables_random.sh
```

**lincRNAs**

```
cd 14-Expression_correlation/cme/

sbatch 16.1-Intergenic-Get_pairs.sh
sbatch 16.2-Intergenic-Create_cis_interaction_table.sh
sbatch 16.3-Intergenic-Create_cis_correlation_table.sh
sbatch 16.4-Intergenic-Join_cis_correlation_tables.sh
sbatch 16.5-Intergenic-Create_cis_correlation_table_random.sh
sbatch 16.6-Intergenic-Join_cis_correlation_tables_random.sh
```
<br />


## Authors

* **Pascual Villalba-Bermell** - *Initial work* - [pasviber](https://github.com/pasviber) (pascual.villalba@csic.es) at [ncRNA-lab](https://github.com/ncRNA-lab).

<br />


## Citation

* <div align="justify"> Villalba-Bermell P., Marquez-Molins J. and Gomez G. Identification, characterization and transcriptional analysis of long non-coding RNAs in Cucurbits. BioRxiv [Preprint]. January 15, 2024. Available from: https://doi.org/10.1101/2024.01.12.575433. </div>

