## Usage

### STEP 0: Select RNA-seq samples

<div align="justify"> We select all RNA-seq samples from Sequence Read Archive database (SRA) and from a particular species. To this end, we use the <a href="https://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/">E-utilities</a>. <br /><br /></div>

```
cd 01-Sample_processing_and_selection/cme/
sbatch 00-Samples.sh
```

<div align="justify"> As a result, we obtain a list of RNA-seq sample accessions and a metadata table with all the information by RNA-seq sample accession.</div>

<br />


### STEP 1: Download RNA-seq samples

<div align="justify"> All raw RNA-seq samples were downloaded from Sequence Read Archive database (SRA) using <a href="https://github.com/ncbi/sra-tools">SRA Toolkit</a>. <br /><br /></div>

```
cd 01-Sample_processing_and_selection/cme/
sbatch 01-Download.sh
```


### STEP 2: Quality control RNA-seq samples

<div align="justify"> All raw RNA-seq samples were trimmed using <a href="https://github.com/OpenGene/fastp">Fastp</a>. <br /><br /></div>

```
cd 01-Sample_processing_and_selection/cme/
sbatch 02-Trimming.sh
```

<div align="justify"> To eliminate possible adapters, the fasta files of the <a href="https://github.com/usadellab/Trimmomatic/tree/main/adapters">Trimmomatic</a> program were used, but other fasta files can be used. It is even possible to gather all adapters in the same file and use a single fasta file. </div>

<div align="justify"> <br />As a summary, a table is generated with the number of reads for each sample. </div>

<br />


### STEP 3: Select strand-specific RNA-seq samples

<div align="justify"> We deduced whether the library is strand-specific or not, and select the strand-specific samples using <a href="https://github.com/COMBINE-lab/salmon">Salmon</a>. <br /><br /></div>

```
cd 01-Sample_processing_and_selection/cme/
sbatch 03-Select_strand-specific_libraries.sh
```

<div align="justify"> As a summary, a table is generated with the library type information for each sample. </div>

<br />


### STEP 4: Mapping

<div align="justify"> We mapped the clean data to the reference genome. To this end, we use <a href="https://github.com/DaehwanKimLab/hisat2">HISAT2</a>. <br /><br /></div>

```
cd 02-Mapping/cme/
sbatch 04-Mapping.sh
```

<div align="justify"> The use of a different number of threads in the HISAT2 program may affect the mapping percentage. However, this variation is very small and has a negligible effect on the final assembly. As this is a pipeline under development, a change of the mapping tool is considered in the future if this problem is not solved. </div>

<br />


### STEP 5: Assembly

<div align="justify"> Bam files generated in the previous step are assembled using <a href="https://github.com/gpertea/stringtie">StringTie2</a>. <br /><br /></div>

```
cd 03-Assembly/cme/
sbatch 05-Assembly.sh
```

<div align="justify"> <br />As a result, an assembly (GTF file) is generated by specie. </div>

<br />


### STEP 6: Transcript annotation

<div align="justify"> Assembled transcripts are classified by their genomic position relative to the protein-coding genes (PCGs) using <a href="https://github.com/gpertea/gffcompare">GffCompare</a>. <br /><br /></div>

```
cd 04-Transcript_annotation/cme/
sbatch 06-Transcript_annotation.sh
```

<div align="justify"> <br />As a result, a class code is assigned to each transcript. </div>

<br />


### STEP 7: LncRNA prediction

<div align="justify"> Assembled and positionally annotated transcripts (GTF file by species) were used to predict potential lncRNAs using several programs and databases (more details in the paper). <br /><br /></div>

```
cd 05-LncRNAs_prediction/cme/
sbatch 07-LncRNAs_prediction.sh
```

<div align="justify"> <br />As a result, we obtain a potential lncRNAs database by species in FASTA, GTF and TSV format. </div>

<br />


### STEP 8: Quantification

<div align="justify"> Bam files generated in the previous step are assembled using <a href="https://github.com/gpertea/stringtie">StringTie2</a>. <br /><br /></div>

```
cd 06-Quantification/cme/
sbatch 08-Quantification_ALL.sh
sbatch 08-Quantification_LncRNAs.sh
sbatch 08-Quantification_Genes.sh
```


### STEP 9: Get random intergenic regions

<div align="justify"> Bam files generated in the previous step are assembled using <a href="https://github.com/gpertea/stringtie">StringTie2</a>. <br /><br /></div>

```
cd 07-Get_intergenic_regions/cme/
sbatch 09-Random_intergenic_regions.sh
```


### STEP 10: Search for TEs and genomic repeats

<div align="justify"> Bam files generated in the previous step are assembled using <a href="https://github.com/gpertea/stringtie">StringTie2</a>. <br /><br /></div>

```
cd 08-TEs_and_genomic_repeats/cme/
sbatch 10.1-RepeatModeler.sh
sbatch 10.2-RepeatMasker.sh
sbatch 10.3-Intersection_with_PCGs_LncRNAs_and_IR.sh
```


### STEP 11: Molecular properties comparison

<div align="justify"> Molecular properties comparison. Bam files generated in the previous step are assembled using <a href="https://github.com/gpertea/stringtie">StringTie2</a>. <br /><br /></div>

```
cd 09-Comparison_PCGs_and_lncRNAs/cme/
sbatch 11-Comparison_PCGs_and_lncRNAs.sh
```


### STEP 12: Genomic distribution

<div align="justify"> Bam files generated in the previous step are assembled using <a href="https://github.com/gpertea/stringtie">StringTie2</a>. <br /><br /></div>

```
cd 10-Genomic_distribution/cme/
sbatch 06-Transcript_annotation.sh
```


### STEP 13: Comparative genomics

<div align="justify"> Bam files generated in the previous step are assembled using <a href="https://github.com/gpertea/stringtie">StringTie2</a>. <br /><br /></div>

```
cd 11-Comparative_genomics/Sequence_level/
sbatch 13-Sequence_level_analysis.sh

cd 11-Comparative_genomics/Positional_level/
sbatch 13.1-LncRNA_and_PCG_ids_list.sh
sbatch 13.2-Orthologs_identification.sh
sbatch 13.3-Find_syntenic_lncRNAs.sh
sbatch 13.4-Classify_lncRNAs_into_families.sh
sbatch 13.5-Create_figures_and_tables.sh

cd 11-Comparative_genomics/Motif_level/
sbatch 13.1-Select_lncRNAs.sh
sbatch 13.2-Prepare_positional_conserved_family_files.sh
sbatch 13.3-Motif_finder-MEME.sh
sbatch 13.4-Motif_enrichment-GOMO.sh
sbatch 13.5-Create_summary_tables.sh
sbatch 13.6-Create_figures_and_tables.sh
```


### STEP 14: Tissue-specificity analysis

<div align="justify"> Bam files generated in the previous step are assembled using <a href="https://github.com/gpertea/stringtie">StringTie2</a>. <br /><br /></div>

```
cd 12-Tissue-specificity/cme/
sbatch 14-Tissue-specificity_analysis.sh
```


### STEP 15: Differential expression analysis

<div align="justify"> Bam files generated in the previous step are assembled using <a href="https://github.com/gpertea/stringtie">StringTie2</a>. <br /><br /></div>

```
cd 13-DEA/cme/
sbatch 15-DEA.sh
```


- [FastQC](https://github.com/s-andrews/FastQC) v.0.11.9
- [Multiqc](https://github.com/MultiQC/MultiQC) v.1.11

- [RSEM](https://github.com/deweylab/RSEM) v.1.3.3


- [Bedtools](https://bedtools.readthedocs.io/en/latest/) v.2.27.1
- [RepeatModeler](https://www.repeatmasker.org/RepeatModeler/) v.2.0.3
- [RepeatMasker](https://www.repeatmasker.org/RepeatMasker/) v.4.1.3-p1
- [OrthoFinder](https://github.com/davidemms/OrthoFinder) v.2.5.4
- [MEME](https://meme-suite.org/meme/) v.5.5.1
- [Tspex](https://apcamargo.github.io/tspex/) (tissue-specificity calculator tool)


- [Swissprot](https://www.uniprot.org/help/downloads)
- [Pfam](https://www.ebi.ac.uk/interpro/download/pfam/)
- [RNAcentral](https://rnacentral.org/)
- [miRBase](https://mirbase.org/)
- [PmiREN](https://www.pmiren.com/)
- [CANTATAdb](http://cantata.amu.edu.pl/)
- [PLncDB](https://www.tobaccodb.org/plncdb/)
- [GreeNC](http://greenc.sequentiabiotech.com/wiki2/Main_Page)

<br />

## Authors

* **Pascual Villalba-Bermell** - *Initial work* - [pasviber](https://github.com/pasviber) (pascual.villalba@csic.es) at [ncRNA-lab](https://github.com/ncRNA-lab).

<br />

